lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,748 @@
+    // of a binary op
+    genBinOp(pred, op, lhs, rhs, 1);
+  }
+
+  // Muxes generally convert to an ite statement
+  void visit(comb::MuxOp op) {
+    // Extract predicate, true and false values
+    Value pred = op.getCond();
+    Value tval = op.getTrueValue();
+    Value fval = op.getFalseValue();
+
+    // We assume that both tval and fval have the same width
+    // This width should be the same as the output width
+    int64_t w = requireSort(op.getType());
+
+    // Generate the ite instruction
+    genIte(op, pred, tval, fval, w);
+  }
+
+  // Assertions are negated then converted to a btor2 bad instruction
+  void visit(sv::AssertOp op) {
+    // Expression is what we will try to invert for our assertion
+    Value expr = op.getExpression();
+
+    // This sort is for assertion inversion and potential implies
+    genSort(bitvecStr, 1);
+
+    // Check for an overaching enable
+    // In our case the sv.if operation will probably only be used when
+    // conditioning an sv.assert on an enable signal. This means that
+    // its condition is probably used to imply our assertion
+    if (auto ifop = dyn_cast<sv::IfOp>(((Operation *)op)->getParentOp())) {
+      Value en = ifop.getOperand();
+
+      // Generate the implication
+      genImplies(ifop, en, expr);
+
+      // Generate the implies inversion
+      genUnaryOp(op, ifop, notStr, 1);
+    } else {
+      // Generate the expression inversion
+      genUnaryOp(op, expr, notStr, 1);
+    }
+
+    // Genrate the bad btor2 intruction
+    genBad(op);
+  }
+  // Assumptions are converted to a btor2 constraint instruction
+  void visit(sv::AssumeOp op) {
+    // Extract the expression that we want our constraint to be about
+    Value expr = op.getExpression();
+    genConstraint(expr);
+  }
+
+  // Firrtl registers generate a state instruction
+  // The final update is also used to generate a set of next btor
+  // instructions
+  void visit(seq::FirRegOp reg) {
+    // Start by retrieving the register's name and width
+    StringRef regName = reg.getName();
+    int64_t w = requireSort(reg.getType());
+
+    // Generate state instruction (represents the register declaration)
+    genState(reg, w, regName);
+
+    // Record the operation for future `next` instruction generation
+    // This is required to model transitions between states (i.e. how a
+    // register's value evolves over time)
+    regOps.push_back(reg);
+  }
+};
+} // end anonymous namespace
+
+void ConvertHWToBTOR2Pass::runOnOperation() {
+  // Btor2 does not have the concept of modules or module
+  // hierarchies, so we assume that no nested modules exist at this point.
+  // This greatly simplifies translation.
+  getOperation().walk([&](hw::HWModuleOp module) {
+    // Start by extracting the inputs and generating appropriate instructions
+    for (auto &port : module.getPortList()) {
+      visit(port);
+    }
+
+    // Visit all of the operations in our module
+    module.walk([&](Operation *op) {
```

Comment:
I agreed with the earlier comment about inheriting the dialect visitors, but I don't think that is being done here is it?

I would expect to see something inherit, e.g., CombinationalVisitor like this: https://github.com/llvm/circt/blob/3fea5dfca33cb0c4dc8ffdc96f9bcd43f4906634/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L2041

And during this walk, instead of the big TypeSwitch, call dispatchCombinationalVistor: https://github.com/llvm/circt/blob/3fea5dfca33cb0c4dc8ffdc96f9bcd43f4906634/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L2535

There are other Visitors for SV, etc. with a similar pattern, and you can inherit from the ones you'd need. I think if you refactor this a bit to use the Visitors, you should be able to keep all your `visit` methods and just rename them a bit. The visitors also give you default unhandled operation methods, which you can override if needed.

The ExportVerilog pass in those links has some other examples, and please let us know if you have any questions.

Reply:
Yeah, I wasn't really sure that I understood what I should inherit here and was actually hoping that someone would add additional feedback for that, thanks for the info, I see how this can definitely make the code a lot cleaner!